File: spark16/src/main/java/scalassssss/SparkHbase.scala
package scalassssss

import org.apache.hadoop.hbase.{HBaseConfiguration, TableName}
import org.apache.hadoop.hbase.client.{HTable, Put}
import org.apache.hadoop.hbase.util.Bytes
import org.apache.spark.{SparkConf, SparkContext}

/**
  * Created by jk on 2016/7/26.
  */
object SparkHbase {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("SparkHbase").setMaster("local")
    val sc = new SparkContext(conf)
    val readFile = sc.textFile("/path/to/file").map(x => x.split(","))
    val tableName = "table"

    readFile.foreachPartition{
      x=> {
        val myConf = HBaseConfiguration.create()
        myConf.set("hbase.zookeeper.quorum", "web102,web101,web100")
        myConf.set("hbase.zookeeper.property.clientPort", "2181")
        myConf.set("hbase.defaults.for.version.skip", "true")
        val myTable = new HTable(myConf, TableName.valueOf(tableName))
        myTable.setAutoFlush(false, false)//关键点1
        myTable.setWriteBufferSize(3*1024*1024)//关键点2
        x.foreach { y => {
          println(y(0) + ":::" + y(1))
          val p = new Put(Bytes.toBytes(y(0)))
          p.add("Family".getBytes, "qualifier".getBytes, Bytes.toBytes(y(1)))
          myTable.put(p)
        }
        }
        myTable.flushCommits()//关键点3
      }
    }
  }

}

File: spark16/src/main/java/javasssss/SparkHBase.java
package javasssss;

import org.apache.commons.io.IOUtils;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hbase.HBaseConfiguration;
import org.apache.hadoop.hbase.KeyValue;
import org.apache.hadoop.hbase.client.*;
import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
import org.apache.hadoop.hbase.mapreduce.TableInputFormat;
import org.apache.hadoop.hbase.util.Base64;
import org.apache.hadoop.hbase.util.Bytes;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.util.List;

public class SparkHBase {

	public static void main(String[] args) throws Exception {
		SparkConf sparkConf = new SparkConf();
        sparkConf.setMaster("spark://centos.host1:7077");  
        sparkConf.setAppName("Spark HBase");  
        JavaSparkContext context = new JavaSparkContext(sparkConf);  
  
        Configuration conf = HBaseConfiguration.create();  
        conf.set("hbase.zookeeper.property.clientPort", "2181");  
        conf.set("hbase.zookeeper.quorum", "centos.host1");  
        conf.set("hbase.master", "centos.host1:60000");  
        
        //Put操作
        HTable table = new HTable(conf, "user");  
        Put put = new Put(Bytes.toBytes("row6"));  
        put.add(Bytes.toBytes("basic"), Bytes.toBytes("name"), Bytes.toBytes("value6"));  
        table.put(put);  
        table.flushCommits();  
        
        //Delete操作
        Delete delete = new Delete(Bytes.toBytes("row1"));
        table.delete(delete);
        
        table.close();  
        
        //Scan操作
        Scan scan = new Scan();  
        scan.setStartRow(Bytes.toBytes("0120140722"));  
        scan.setStopRow(Bytes.toBytes("1620140728"));  
        scan.addFamily(Bytes.toBytes("basic"));  
        scan.addColumn(Bytes.toBytes("basic"), Bytes.toBytes("name"));  
  
        String tableName = "user";  
        conf.set(TableInputFormat.INPUT_TABLE, tableName);  
  
        ByteArrayOutputStream out = new ByteArrayOutputStream();  
        DataOutputStream dos = new DataOutputStream(out);  
//        scan.write(dos);
        String scanStr = Base64.encodeBytes(out.toByteArray());  
        IOUtils.closeQuietly(dos);
        IOUtils.closeQuietly(out);  
        //高版本可以用如下方式：  
        //ClientProtos.Scan proto = ProtobufUtil.toScan(scan);  
        //String scanStr = Base64.encodeBytes(proto.toByteArray());  
        conf.set(TableInputFormat.SCAN, scanStr);  
  
        JavaPairRDD<ImmutableBytesWritable, Result> hBaseRDD = context  
                .newAPIHadoopRDD(conf, TableInputFormat.class,  
                        ImmutableBytesWritable.class, Result.class);
            
//            hBaseRDD.saveAsHadoopDataset();
            
            Long count = hBaseRDD.count();
        System.out.println("count: " + count);  
  
        List<Tuple2<ImmutableBytesWritable, Result>> tuples = hBaseRDD  
                .take(count.intValue());  
        for (int i = 0, len = count.intValue(); i < len; i++) {  
            Result result = tuples.get(i)._2();  
            KeyValue[] kvs = result.raw();  
            for (KeyValue kv : kvs) {  
                System.out.println("rowkey:" + new String(kv.getRow()) + " cf:"  
                        + new String(kv.getFamily()) + " column:"  
                        + new String(kv.getQualifier()) + " value:"  
                        + new String(kv.getValue()));  
            }  
        }  
    }  
}